lib/os/queued_operation.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys/queued_operation.h>
+
+static inline int op_get_priority(const struct queued_operation *op)
+{
+	return (s8_t)(op->notify.flags >> QUEUED_OPERATION_PRIORITY_POS);
+}
+
+static inline int op_set_priority(struct queued_operation *op,
+				  int priority)
+{
+	s8_t prio = (s8_t)priority;
+	u32_t mask = (QUEUED_OPERATION_PRIORITY_MASK << QUEUED_OPERATION_PRIORITY_POS);
+
+	if (prio != priority) {
+		return -EINVAL;
+	}
+
+	op->notify.flags = (op->notify.flags & ~mask)
+			   | (mask & (prio << QUEUED_OPERATION_PRIORITY_POS));
+
+	return 0;
+}
+
+static inline bool can_start(/*const*/ struct queued_operation_manager *mgr)
+{
+	/* OK if not busy, not finalizing, and an op is available. */
+	return (mgr->current == NULL)
+	       && !mgr->finalizing
+	       && !sys_slist_is_empty(&mgr->operations);
+}
+
+static void select_next_and_unlock(struct queued_operation_manager *mgr,
+				   k_spinlock_key_t key)
+{
+	struct queued_operation *op = NULL;
+
+	/* Track whether the manager is idle, so we only send
+	 * notification of entry to idle once.
+	 */
+	bool in_idle = (mgr->current == NULL);
+
+	do {
+		/* Can't select a new operation if one is active. */
+		if (mgr->current != NULL) {
+			break;
+		}
+
+		/* Can't select a new operation until the previous one
+		 * completes finalization, lest we pick something that
+		 * has too low a priority.
+		 */
+		if (mgr->finalizing) {
+			break;
+		}
+
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->current = op;
+		} else {
+			op = NULL;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+
+		/* Only notify the manager if there's an operation, or
+		 * if it is to transition to idle.
+		 */
+		if ((op != NULL) || !in_idle) {
+			mgr->vtable->process(mgr, op);
+		}
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
+
+	k_spin_unlock(&mgr->lock, key);
+}
+
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority)
+{
+	int rv = 0;
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(mgr->vtable != NULL);
+	__ASSERT_NO_MSG(mgr->vtable->process != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	/* Validation is optional; if present, use it. */
+	if (mgr->vtable->validate) {
+		rv = mgr->vtable->validate(mgr, op);
+	}
+
+	/* Set the priority, checking whether it's in range. */
+	if (rv == 0) {
+		rv = op_set_priority(op, priority);
+	}
+
+	/* Reject callback notifications without translation
+	 * function.
+	 */
+	if ((rv == 0)
+	    && async_notify_uses_callback(&op->notify)
+	    && (mgr->vtable->callback == NULL)) {
+		rv = -ENOTSUP;
+	}
+
+	if (rv < 0) {
+		return rv;
+	}
+
```

Comment:
probably in majority of cases there will single operation only, it would be good to optimized this. I've been thinking about two things:
- using only `current` to indicate that module is busy, finalizing stage can be encoded by using invalid address (e.g. `UINTPTR_MAX`)
- when it is used then process can be started immediately and locklessy (note that for that atomic.h must be extended: #22680 :
```
if (atomic_cas((atomic_t *)&mgr->current, (atomic_t)NULL, (atomic_t)op)) {
	mgr->vtable->process(mgr, op);
	return rv;
}
```

I tried it (and adapt getting next op) and measured time which went from 5.5us to 3.9us on nrf52.
I was trying if it is feasible (https://github.com/nordic-krch/zephyr/commit/47371da4ec69fec96ce93a7e3b06e5c440740379)

Reply:
OK, I've addressed all the concerns except this one.  (Aside: I think you have the wrong link for an extension to atomic.h.)

I'm not taking this suggestion, as I think it adds significant complexity (magic non-pointer values stored in pointers; a new synchronization mechanism) with little real value (1.5 us per submitted operation).

Reply:
> I'm not taking this suggestion, as I think it adds significant complexity 

Here i disagree because:
- one variable to keep busy state which is simpler
- selecting next to process is simplified
- speed up (you say 1.5us, I say 30%)
- a bit less rom (150 bytes less)